clippy_lints/src/no_mangle_with_rust_abi.rs

```diff
@@ -0,0 +1,60 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use rustc_errors::Applicability;
+use rustc_hir::{Item, ItemKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_target::spec::abi::Abi;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for Rust ABI functions with the `#[no_mangle]` attribute.
+    ///
+    /// ### Why is this bad?
+    /// The Rust ABI is not stable, but in many simple cases matches 
+    /// enough with the C ABI that it is possible to forget to add 
+    /// `extern "C"` to a function called from C. Changes to the 
+    /// Rust ABI can break this at any point.
+    ///
+    /// ### Example
+    /// ```rust
+    ///  #[no_mangle]
+    ///  fn example(arg_one: u32, arg_two: usize) {}
+    /// ```
+    ///
+    /// Use instead:
+    /// ```rust
+    ///  #[no_mangle]
+    ///  extern "C" fn example(arg_one: u32, arg_two: usize) {}
+    /// ```
+    #[clippy::version = "1.69.0"]
+    pub NO_MANGLE_WITH_RUST_ABI,
+    suspicious,
+    "convert Rust ABI functions to C ABI"
+}
+declare_lint_pass!(NoMangleWithRustAbi => [NO_MANGLE_WITH_RUST_ABI]);
+
+impl<'tcx> LateLintPass<'tcx> for NoMangleWithRustAbi {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if let ItemKind::Fn(fn_sig, _, _) = &item.kind {
+            let attrs = cx.tcx.hir().attrs(item.hir_id());
+            for attr in attrs {
+                if let Some(ident) = attr.ident()
+                && ident.name.as_str() == "no_mangle" 
+                && fn_sig.header.abi == Abi::Rust {
+                let mut applicability = Applicability::MachineApplicable;
+                let span = fn_sig.span;
+                span_lint_and_sugg(
+                    cx,
+                    NO_MANGLE_WITH_RUST_ABI,
+                    span,
+                    &format!("attribute #[no_mangle] set on a Rust ABI function"),
+                    "try",
+                    format!(r#"extern "C" {}"#, snippet_with_applicability(cx, span, "..", &mut applicability)),
```

Comment:
@llogiq I'm not sure about the best way to accomplish this. I have the entire `span` available here - `pub fn example()`, which needs to be modified to `pub extern "C" fn example()`. Is there an elegant way to do this that doesn't involve string splitting?

Reply:
Ended up using `String::split_once("fn")`